petastorm/pytorch.py

```diff
@@ -367,3 +367,123 @@ def __enter__(self):
         self.reader.stop()
         self.reader.join()
+
+
+class InMemBatchedDataLoader(LoaderBase):
+    """
+    Same as BatchedDataLoader except it only loads upto capacity rows into memory.
+    """
+
+    def __init__(self, reader, batch_size=1,
+                 transform_fn=None,
+                 num_epochs=1,
+                 seed=0,
+                 capacity=1024,
+                 shuffle=False):
+        """
+        Initializes a data loader object.
+        This class does not support special types that are not supported in PyTorch (decimal/string).
+        :param reader: petastorm Reader instance
+        :param batch_size: the number of items to return per batch; factored into the len() of this reader
+        :param transform_fn: an optional callable to convert batches from the reader to PyTorch tensors
+        :param num_epochs: number of epochs.
+        :param seed: random seed used to shuffle.
+        :param capacity: number of rows to be loaded into memory.
+        :param shuffle: If ``True``, indices will be shuffled in every epoch.
+        """
+        super(InMemBatchedDataLoader, self).__init__()
+        self.reader = reader
+        self.batch_size = batch_size
+        self.transform_fn = transform_fn or torch.as_tensor
+        self.num_epochs = num_epochs
+        self.seed = seed
+        self.capacity = capacity
+        self.shuffle = shuffle
+        self.size = 0
+
+    def _load_rows_into_mem(self):
+        n_rows = 0
+        buffer_full = False
+        self.buffer = None
+        self.keys = None
+
+        for row in self.reader:
+            if buffer_full:
+                break
+            # Default collate does not work nicely on namedtuples and treat them as lists
+            # Using dict will result in the yielded structures being dicts as well
+            row_as_dict = row._asdict()
+
+            # Promote some types that are incompatible with pytorch to be pytorch friendly.
+            _sanitize_pytorch_types(row_as_dict)
+
+            for k, v in row_as_dict.items():
+                if not self.reader.batched_output:
+                    row_as_dict[k] = self.transform_fn([v])
+                else:
+                    row_as_dict[k] = self.transform_fn(v)
+
+            if not self.keys:
+                self.keys = row_as_dict.keys()
+
+            # Add rows to buffer
+            items = list(row_as_dict.values())
+            expected_rows = n_rows + len(items[0])
+            last_row = len(items[0])
+
+            if self.capacity <= expected_rows:
+                buffer_full = True
+                last_row = self.capacity-n_rows
+                expected_rows = self.capacity
+            if self.buffer is None:
+                # Initialize buffer as a list of empty tensors
+                self.buffer = []
+                for v in items:
+                    self.buffer.append(torch.empty((self.capacity,) + v.shape[1:], dtype=v.dtype, device=v.device))
```

Comment:
Can you please clarify what does the `+v.shape[1:]` term purpose?

Reply:
We want to allocate empty tensor with exact dimension of input rows.
v.shape[0] is row group size, while v.shape[1:] is possible dimensions of input row data.
This is same as code in shuffling buffer:  https://github.com/uber/petastorm/blob/7f37e8dde6ff1b13f055d22a6289e2de8bb5d473/petastorm/reader_impl/pytorch_shuffling_buffer.py#L269